src/compiler/watchPublic.ts

```diff
@@ -434,15 +435,22 @@ namespace ts {
             const hasInvalidatedResolution = resolutionCache.createHasInvalidatedResolution(userProvidedResolution || changesAffectResolution);
             if (isProgramUptoDate(getCurrentProgram(), rootFileNames, compilerOptions, getSourceVersion, fileExists, hasInvalidatedResolution, hasChangedAutomaticTypeDirectiveNames, getParsedCommandLine, projectReferences)) {
                 if (hasChangedConfigFileParsingErrors) {
+                    if (reportFileChangeDetectedOnCreateProgram) {
+                        reportWatchDiagnostic(Diagnostics.File_change_detected_Starting_incremental_compilation);

```

Comment:
How much work is done between where this used to be reported and here?  I'm concerned that it will look like the compiler is frozen if it reacts "late" to a file change.

Reply:
That's a good question; my skimming says that `isProgramUptoDate` will:

- Compare a few arrays for equality
- Check for not-up-to-date source files, which appears to be checking a number of caches
- Check for files that the program thought were missing on disk
- Compare compiler options
- Compare tsconfig contents by text for position changes for diagnostics

I'm not sure that it's a huge amount of work.

I could try and have it unconditionally run `afterProgramCreate`, but based on #37266, it seems like this was an intentional change (but, hard to tell entirely, given it's one part of a lot of other tweaks). I'll try it and see if anything breaks catastrophically, but I don't know exactly how to double check that I don't undo a perf fix.

Reply:
I think we convinced ourselves that it should be cheap.  I'm fine bugging this into shape if we get complaints.